### PR TITLE
Feat / XEMM - Add more Price Sources for Base+Quote Conversion rates

### PR DESCRIPTION
**Pull request by TheHolyRoger**
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Depends on PRs #3128 and #3123
This PR must be rebased after those PRs are merged. ⚠️ 

Adds more price sources using order books for conversation rates in Cross Exchange Market Making

### CODE REVIEW

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 1]
```diff
@@ -1073,10 +1142,14 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
...
-
-            object quote_asset_amount = maker_market.c_get_balance(market_pair.maker.quote_asset) if is_buy else \
-                taker_market.c_get_balance(market_pair.taker.quote_asset)
-            object base_asset_amount = taker_market.c_get_balance(market_pair.taker.base_asset) if is_buy else \
+            object base_rate = self.market_conversion_rate('base')
+            object quote_rate = self.market_conversion_rate('quote')
+
+            object quote_asset_amount = maker_market.c_get_balance(market_pair.maker.quote_asset) \
+                if is_buy else \
+                taker_market.c_get_balance(market_pair.taker.quote_asset) * quote_rate
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1150]: Conversion calculation issue.

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 2]
```diff
@@ -1073,10 +1142,14 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
...
+            object base_asset_amount = taker_market.c_get_balance(market_pair.taker.base_asset) * base_rate \
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1151]: Conversion calculation issue.

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 3]
```diff
@@ -1137,15 +1216,15 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
                         True,
                         bid_size
                     )
-                    effective_hedging_price_adjusted = effective_hedging_price * self.market_conversion_rate()
+                    effective_hedging_price_adjusted = effective_hedging_price / self.market_conversion_rate()
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1219]: Conversion calculation issue.

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 4]
```diff
@@ -1137,15 +1216,15 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
...
                     if self._logging_options & self.OPTION_LOG_CREATE_ORDER:
                         self.log_with_clock(
                             logging.INFO,
                             f"({market_pair.maker.trading_pair}) Creating limit bid order for "
                             f"{bid_size} {market_pair.maker.base_asset} at "
                             f"{bid_price} {market_pair.maker.quote_asset}. "
-                            f"Current hedging price: {effective_hedging_price} {market_pair.taker.quote_asset} "
-                            f"(Rate adjusted: {effective_hedging_price_adjusted:.2f} {market_pair.taker.quote_asset})."
+                            f"Current hedging price: {effective_hedging_price:.8f} {market_pair.maker.quote_asset} "
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1226]: Fixed wrong asset in log.

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 5]
```diff
@@ -1137,15 +1216,15 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
...
+                            f"(Rate adjusted: {effective_hedging_price_adjusted:.8f} {market_pair.taker.quote_asset})."
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1227]: When working with some tokens, 2 decimal places isn't enough

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 6]
```diff
@@ -1176,15 +1255,15 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
                         False,
                         ask_size
                     )
-                    effective_hedging_price_adjusted = effective_hedging_price
+                    effective_hedging_price_adjusted = effective_hedging_price / self.market_conversion_rate()
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1258]: Conversion calculation issue.

File: hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx [hunk 7]
```diff
@@ -1176,15 +1255,15 @@ cdef class CrossExchangeMarketMakingStrategy(StrategyBase):
...
                     if self._logging_options & self.OPTION_LOG_CREATE_ORDER:
                         self.log_with_clock(
                             logging.INFO,
                             f"({market_pair.maker.trading_pair}) Creating limit ask order for "
                             f"{ask_size} {market_pair.maker.base_asset} at "
                             f"{ask_price} {market_pair.maker.quote_asset}. "
-                            f"Current hedging price: {effective_hedging_price} {market_pair.maker.quote_asset} "
-                            f"(Rate adjusted: {effective_hedging_price_adjusted:.2f} {market_pair.maker.quote_asset})."
+                            f"Current hedging price: {effective_hedging_price:.8f} {market_pair.maker.quote_asset} "
+                            f"(Rate adjusted: {effective_hedging_price_adjusted:.8f} {market_pair.taker.quote_asset})."
```
Comment (TheHolyRoger) [inline hummingbot/strategy/cross_exchange_market_making/cross_exchange_market_making.pyx:1266]: Same 2 decimal place issue + wrong asset name in log

**Review by TheHolyRoger — COMMENTED** [review]
@Nullably @phbrgnomo 
I found some maths issues in XEMM which I've also fixed in this PR, affects+fixes rate oracle conversions too, but it's worth you guys checking I've not made a mistake?

**Comment by phbrgnomo** [comment]
As mentioned on the PR #2865, are are goig to employ different approaches to this.

Thank you again for this contribution, helped us to better understand how to do it, but we will close it for now.

**Pull request closed without merging**